src/ucp/core/ucp_mm.c

```diff
@@ -1305,3 +1412,349 @@ ucs_status_t ucp_mem_reg_md_map_update(ucp_context_h context)
+    const void *next_tl_md_p, *end;
+    size_t tl_mkey_data_size;
+    size_t tl_mkey_size, global_id_size;
+    const void *tl_mkey_buf, *global_id_buf;
+    ucp_md_map_t md_map;
+
+    tl_mkey_data_size = ucp_memh_info_size_unpack(&p);
+    end               = UCS_PTR_BYTE_OFFSET(*start_p, tl_mkey_data_size);
+    ucs_assert(tl_mkey_data_size != 0);
+
+    tl_mkey_size = *ucs_serialize_next(&p, uint8_t);
+    ucs_assert(tl_mkey_size != 0);
+
+    tl_mkey_buf = ucs_serialize_next_raw(&p, void, tl_mkey_size);
+
+    global_id_size = *ucs_serialize_next(&p, uint8_t);
+    ucs_assert(global_id_size != 0);
+
+    global_id_buf = ucs_serialize_next_raw(&p, void, global_id_size);
+
+    /* Get local MD indices which corresponds to the remote ones */
+    md_map = ucp_find_md_map(context, global_id_buf, global_id_size);
+
+    next_tl_md_p = end;
+    ucs_assertv(p <= next_tl_md_p, "p=%p, next_tl_md_p=%p", p, next_tl_md_p);
+
+    *start_p       = next_tl_md_p;
+    *tl_mkey_buf_p = tl_mkey_buf;
+    *md_map_p      = md_map;
+}
+
+static ucs_status_t
+ucp_memh_import_attach(ucp_context_h context, ucp_mem_h memh,
+                       ucp_memh_import_attach_params_t *attach_params_array,
+                       unsigned attach_params_num)
+{
+    ucp_md_index_t md_index = UCP_NULL_RESOURCE;
+    unsigned attach_params_iter;
+    const void *tl_mkey_buf;
+    uct_md_mem_attach_params_t attach_params;
+    uct_md_attr_v2_t *md_attr;
+    ucs_status_t status;
+    uct_mem_h uct_memh;
+
+    for (attach_params_iter = 0; attach_params_iter < attach_params_num;
+         ++attach_params_iter) {
+        md_index    = attach_params_array[attach_params_iter].md_index;
+        tl_mkey_buf = attach_params_array[attach_params_iter].tl_mkey_buf;
+        md_attr     = &context->tl_mds[md_index].attr;
+        ucs_assert_always(md_attr->flags & UCT_MD_FLAG_EXPORTED_MKEY);
+
+        if (memh->uct[md_index] != NULL) {
+            continue;
+        }
+
+        attach_params.field_mask = UCT_MD_MEM_ATTACH_FIELD_FLAGS;
+        attach_params.flags      = UCT_MD_MEM_ATTACH_FLAG_HIDE_ERRORS;
+
+        status = uct_md_mem_attach(context->tl_mds[md_index].md, tl_mkey_buf,
+                                   &attach_params, &uct_memh);
+        if (ucs_unlikely(status != UCS_OK)) {
+            /* Don't print an error, because two MDs can have similar global
+             * identifiers, but a memory key was exported on another MD */
+            ucs_trace("failed to attach memory on '%s': %s",
+                      md_attr->component_name, ucs_status_string(status));
+            continue;
+        }
+
+        memh->uct[md_index] = uct_memh;
+
+        ucs_trace("imported address %p length %zu on md[%d]=%s: uct_memh %p",
+                  ucp_memh_address(memh), ucp_memh_length(memh), md_index,
+                  context->tl_mds[md_index].rsc.md_name,
+                  memh->uct[md_index]);
+        memh->md_map |= UCS_BIT(md_index);
```

Comment:
minor: better have it near `memh->uct` assignment, before `ucs_trace`

Reply:
fixed